rmw_implementation/src/functions.cpp

```diff
@@ -182,491 +199,210 @@ extern "C"
+  rmw_ret_t, RMW_RET_ERROR,
+  4, ARG_TYPES(
+    const rmw_node_t *, rcutils_allocator_t *, bool,
+    rmw_names_and_types_t *))
+
+RMW_INTERFACE_FN(rmw_get_service_names_and_types,
+  rmw_ret_t, RMW_RET_ERROR,
+  3, ARG_TYPES(
+    const rmw_node_t *, rcutils_allocator_t *,
+    rmw_names_and_types_t *))
+
+RMW_INTERFACE_FN(rmw_get_node_names,
+  rmw_ret_t, RMW_RET_ERROR,
+  2, ARG_TYPES(const rmw_node_t *, rcutils_string_array_t *))
+
+RMW_INTERFACE_FN(rmw_count_publishers,
+  rmw_ret_t, RMW_RET_ERROR,
+  3, ARG_TYPES(const rmw_node_t *, const char *, size_t *))
+
+RMW_INTERFACE_FN(rmw_count_subscribers,
+  rmw_ret_t, RMW_RET_ERROR,
+  3, ARG_TYPES(const rmw_node_t *, const char *, size_t *))
+
+RMW_INTERFACE_FN(rmw_get_gid_for_publisher,
+  rmw_ret_t, RMW_RET_ERROR,
+  2, ARG_TYPES(const rmw_publisher_t *, rmw_gid_t *))
+
+RMW_INTERFACE_FN(rmw_compare_gids_equal,
+  rmw_ret_t, RMW_RET_ERROR,
+  3, ARG_TYPES(const rmw_gid_t *, const rmw_gid_t *, bool *))
+
+RMW_INTERFACE_FN(rmw_service_server_is_available,
+  rmw_ret_t, RMW_RET_ERROR,
+  3, ARG_TYPES(const rmw_node_t *, const rmw_client_t *, bool *))
+
+#define GET_SYMBOL(x) symbol_ ## x = get_symbol(#x);
 
-rmw_ret_t
-rmw_service_server_is_available(
-  const rmw_node_t * node,
-  const rmw_client_t * client,
-  bool * is_available)
+void prefetch_symbols(void)
 {
-  CALL_SYMBOL(
-    rmw_service_server_is_available, rmw_ret_t, RMW_RET_ERROR,
-    ARG_TYPES(const rmw_node_t *, const rmw_client_t *, bool *),
-    ARG_VALUES(node, client, is_available));
+  // get all symbols to avoid race conditions later since the passed
+  // symbol name is expected to be a std::string which requires allocation
+  GET_SYMBOL(rmw_get_implementation_identifier)
+  GET_SYMBOL(rmw_create_node)
+  GET_SYMBOL(rmw_destroy_node)
+  GET_SYMBOL(rmw_node_get_graph_guard_condition)
+  GET_SYMBOL(rmw_create_publisher)
+  GET_SYMBOL(rmw_destroy_publisher)
+  GET_SYMBOL(rmw_publish)
+  GET_SYMBOL(rmw_create_subscription)
+  GET_SYMBOL(rmw_destroy_subscription)
+  GET_SYMBOL(rmw_take)
+  GET_SYMBOL(rmw_take_with_info)
+  GET_SYMBOL(rmw_create_client)
+  GET_SYMBOL(rmw_destroy_client)
+  GET_SYMBOL(rmw_send_request)
+  GET_SYMBOL(rmw_take_response)
+  GET_SYMBOL(rmw_create_service)
+  GET_SYMBOL(rmw_destroy_service)
+  GET_SYMBOL(rmw_take_request)
+  GET_SYMBOL(rmw_send_response)
+  GET_SYMBOL(rmw_create_guard_condition)
+  GET_SYMBOL(rmw_destroy_guard_condition)
+  GET_SYMBOL(rmw_trigger_guard_condition)
+  GET_SYMBOL(rmw_create_waitset)
+  GET_SYMBOL(rmw_destroy_waitset)
+  GET_SYMBOL(rmw_wait)
+  GET_SYMBOL(rmw_get_topic_names_and_types)
+  GET_SYMBOL(rmw_get_service_names_and_types)
+  GET_SYMBOL(rmw_get_node_names)
+  GET_SYMBOL(rmw_count_publishers)
+  GET_SYMBOL(rmw_count_subscribers)
+  GET_SYMBOL(rmw_get_gid_for_publisher)
+  GET_SYMBOL(rmw_compare_gids_equal)
+  GET_SYMBOL(rmw_service_server_is_available)
 }
 
 void * symbol_rmw_init = nullptr;
 
-rmw_ret_t
-rmw_init(void)
+rmw_ret_t rmw_init()
```

Comment:
nitpick: why this change?

Reply:
In the first round all functions were refactored into macros and in a second step `rmw_init` was extracted again. I committed 7d4b932976c0e1b34e66da1b5e875db45a071ab1 to avoid the diff.